src/mtcp/mtcp_restart.c

```diff
@@ -495,13 +496,90 @@ mysetauxval(char **evp, unsigned long int type, unsigned long int val)
   return -1;
 }
 
-#define PAGESIZE 4096 /* We hardwire this, since we can't make libc calls. */
-// We choose this address based on our observation on CORI that it won't overlap
-// with any other memory segments. We might change this address in the future,
-// for now, it doesn't matters.
-#define vvarStartTmp ((0x40000) - 5*PAGESIZE)
-#define vdsoStartTmp ((0x40000) - 2*PAGESIZE)
+/* We hardwire these, since we can't make libc calls. */
+#define PAGESIZE 4096
+#define ROUNDADDRUP(addr, size) ((addr + size - 1) & ~(size - 1))
+// I think we're forced to use global variables, since
+// unmap_memory_areas_and_restore_vdso uses these variables subsequently.
+uint64_t vvarStartTmp = 0;
+uint64_t vdsoStartTmp = 0;
+
+// This function searches for a free memory region to temporarily remap the vdso
+// and vvar regions to, so that mtcp's vdso and vvar do not overlap with the
+// checkpointed process' vdso and vvar.
+static void
+remap_vdso_and_vvar_regions() {
+  Area area;
+  int rc = 0;
+  uint64_t vvarStart = 0;
+  uint64_t vdsoStart = 0;
+  uint64_t vvarSize = 0;
+  uint64_t vdsoSize = 0;
+  uint64_t prev_addr = 0x10000;
+
+  int mapsfd = mtcp_sys_open2("/proc/self/maps", O_RDONLY);
+  if (mapsfd < 0) {
+    MTCP_PRINTF("error opening /proc/self/maps; errno: %d\n", mtcp_sys_errno);
+    mtcp_abort();
+  }
+
+  while (mtcp_readmapsline(mapsfd, &area)) {
+    if (mtcp_strcmp(area.name, "[vvar]") == 0) {
+      vvarStart = area.addr;
+      vvarSize = area.size;
+    } else if (mtcp_strcmp(area.name, "[vdso]") == 0) {
+      vdsoStart = area.addr;
+      vdsoSize = area.size;
+    }
+
+    if (vvarStartTmp == 0) {
+      // In May 2020, on Cori and elsewhere, vvar is 3 pages and vdso is 2
+      // pages.
+      if (prev_addr + 5 * PAGESIZE <= area.addr) {
```

Comment:
Why are we assuming the `vvar` region's size here? 
The `/proc/maps` should give us the exact size of each region on that platform?

Reply:
hmm, that's true. i think we can try finding the regions and their respective sizes first, then reset the file pointer and look for empty regions to map them too.